diesel/src/expression/functions/mod.rs

```diff
@@ -77,15 +79,153 @@ macro_rules! sql_function_body {
 }
 
 #[macro_export]
-/// Declare a sql function for use in your code. Useful if you have your own SQL functions that
-/// you'd like to use. You can optionally provide a doc string as well. `$struct_name` should just
-/// be any unique name. You will not need to reference it in your code, but it is required due to
-/// the fact that [`concat_idents!` is
-/// useless](https://github.com/rust-lang/rust/issues/29599#issuecomment-153927167).
+#[doc(hidden)]
+#[cfg(not(feature = "with-deprecated"))]
+macro_rules! sql_function_body {
+    ($($args:tt)*) => {
+        compile_error!("You are using a deprecated form of `sql_function!`. \
+        You must enable the `with-deprecated` feature on `diesel`.");
+    }
+}
+
+#[macro_export]
+macro_rules! __diesel_sql_function_body {
+    (
+        meta = ($($meta:tt)*),
+        fn_name = $fn_name:ident,
+        args = ($($arg_name:ident: $arg_type:ty),*),
+        return_type = $return_type:ty,
+    ) => {
+        $($meta)*
+        #[allow(non_camel_case_types)]
+        pub fn $fn_name<$($arg_name),*>($($arg_name: $arg_name),*)
+            -> $fn_name::HelperType<$($arg_name),*>
+        where
+            $($arg_name: $crate::expression::AsExpression<$arg_type>),+
+        {
+            $fn_name::$fn_name {
+                $($arg_name: $arg_name.as_expression()),+
+            }
+        }
+
+        #[doc(hidden)]
+        #[allow(non_camel_case_types, unused_imports)]
+        pub(crate) mod $fn_name {
+            use super::*;
+            use $crate::sql_types::*;
+
+            #[derive(Debug, Clone, Copy, QueryId)]
+            pub struct $fn_name<$($arg_name),*> {
+                $(pub(in super) $arg_name: $arg_name),*
+            }
+
+            pub type HelperType<$($arg_name),*> = $fn_name<$(
+                <$arg_name as $crate::expression::AsExpression<$arg_type>>::Expression
+            ),*>;
+
+            impl<$($arg_name),*> $crate::expression::Expression for $fn_name<$($arg_name),*> where
```

Comment:
Maybe it makes sense to unify those trait implementation with the old ones? (They should be the same module the name of the struct)